File: report/src/main/java/com/vmlens/report/element/LoopRunAndThreadIndex.java
package com.vmlens.report.element;

public class LoopRunAndThreadIndex {

    private final int loopId;
    private final int runId;
    private final int threadIndex;

    public LoopRunAndThreadIndex(int loopId, int runId, int threadIndex) {
        this.loopId = loopId;
        this.runId = runId;
        this.threadIndex = threadIndex;
    }

    public int runId() {
        return runId;
    }

    public int threadIndex() {
        return threadIndex;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        LoopRunAndThreadIndex that = (LoopRunAndThreadIndex) o;
        return loopId == that.loopId && runId == that.runId && threadIndex == that.threadIndex;
    }

    @Override
    public int hashCode() {
        int result = loopId;
        result = 31 * result + runId;
        result = 31 * result + threadIndex;
        return result;
    }

    @Override
    public String toString() {
        return "LoopRunAndThreadIndex{" +
                "loopId=" + loopId +
                ", runId=" + runId +
                ", threadIndex=" + threadIndex +
                '}';
    }
}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/LoopRunAndThreadIndex.scala
package com.anarsoft.race.detection.report.element

class LoopRunAndThreadIndex(val loopId: Int,  val runId: Int,  val threadIndex: Int) {

  override def toString: String = "LoopRunAndThreadIndex{" + "loopId=" + loopId + ", runId=" + runId + ", threadIndex=" + threadIndex + '}'

  private def canEqual(other: Any): Boolean = other.isInstanceOf[LoopRunAndThreadIndex]

  override def equals(other: Any): Boolean = other match {
    case that: LoopRunAndThreadIndex =>
      that.canEqual(this) &&
        loopId == that.loopId &&
        runId == that.runId &&
        threadIndex == that.threadIndex
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(loopId, runId, threadIndex)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
